cpp/src/arrow/flight/sql/odbc/odbcabstraction/odbc_impl/odbc_statement.cc

```diff
@@ -0,0 +1,787 @@
+      return;
+    case SQL_ATTR_ROW_ARRAY_SIZE:
+      m_currentArd->SetHeaderField(SQL_DESC_ARRAY_SIZE, value, bufferSize);
+      return;
+    case SQL_ATTR_ROW_BIND_OFFSET_PTR:
+      m_currentArd->SetHeaderField(SQL_DESC_BIND_OFFSET_PTR, value, bufferSize);
+      return;
+    case SQL_ATTR_ROW_BIND_TYPE:
+      m_currentArd->SetHeaderField(SQL_DESC_BIND_TYPE, value, bufferSize);
+      return;
+    case SQL_ATTR_ROW_OPERATION_PTR:
+      m_currentArd->SetHeaderField(SQL_DESC_ARRAY_STATUS_PTR, value, bufferSize);
+      return;
+    case SQL_ATTR_ROW_STATUS_PTR:
+      m_ird->SetHeaderField(SQL_DESC_ARRAY_STATUS_PTR, value, bufferSize);
+      return;
+    case SQL_ATTR_ROWS_FETCHED_PTR:
+      m_ird->SetHeaderField(SQL_DESC_ROWS_PROCESSED_PTR, value, bufferSize);
+      return;
+
+    case SQL_ATTR_ASYNC_ENABLE:
+#ifdef SQL_ATTR_ASYNC_STMT_EVENT
+    case SQL_ATTR_ASYNC_STMT_EVENT:
+      throw DriverException("Unsupported attribute", "HYC00");
+#endif
+#ifdef SQL_ATTR_ASYNC_STMT_PCALLBACK
+    case SQL_ATTR_ASYNC_STMT_PCALLBACK:
+      throw DriverException("Unsupported attribute", "HYC00");
+#endif
+#ifdef SQL_ATTR_ASYNC_STMT_PCONTEXT
+    case SQL_ATTR_ASYNC_STMT_PCONTEXT:
+      throw DriverException("Unsupported attribute", "HYC00");
+#endif
+    case SQL_ATTR_CONCURRENCY:
+      CheckIfAttributeIsSetToOnlyValidValue(value,
+                                            static_cast<SQLULEN>(SQL_CONCUR_READ_ONLY));
+      return;
+    case SQL_ATTR_CURSOR_SCROLLABLE:
+      CheckIfAttributeIsSetToOnlyValidValue(value,
+                                            static_cast<SQLULEN>(SQL_NONSCROLLABLE));
+      return;
+    case SQL_ATTR_CURSOR_SENSITIVITY:
+      CheckIfAttributeIsSetToOnlyValidValue(value, static_cast<SQLULEN>(SQL_UNSPECIFIED));
+      return;
+    case SQL_ATTR_CURSOR_TYPE:
+      CheckIfAttributeIsSetToOnlyValidValue(
+          value, static_cast<SQLULEN>(SQL_CURSOR_FORWARD_ONLY));
+      return;
+    case SQL_ATTR_ENABLE_AUTO_IPD:
+      CheckIfAttributeIsSetToOnlyValidValue(value, static_cast<SQLULEN>(SQL_FALSE));
+      return;
+    case SQL_ATTR_FETCH_BOOKMARK_PTR:
+      if (value != NULL) {
+        throw DriverException("Optional feature not implemented", "HYC00");
+      }
+      return;
+    case SQL_ATTR_KEYSET_SIZE:
+      CheckIfAttributeIsSetToOnlyValidValue(value, static_cast<SQLULEN>(0));
+      return;
+    case SQL_ATTR_ROW_NUMBER:
+      throw DriverException("Cannot set read-only attribute", "HY092");
+    case SQL_ATTR_SIMULATE_CURSOR:
+      CheckIfAttributeIsSetToOnlyValidValue(value, static_cast<SQLULEN>(SQL_SC_UNIQUE));
+      return;
+    case SQL_ATTR_USE_BOOKMARKS:
+      CheckIfAttributeIsSetToOnlyValidValue(value, static_cast<SQLULEN>(SQL_UB_OFF));
+      return;
+    case SQL_ATTR_RETRIEVE_DATA:
+      CheckIfAttributeIsSetToOnlyValidValue(value, static_cast<SQLULEN>(SQL_TRUE));
+      return;
+    case SQL_ROWSET_SIZE:
+      SetAttribute(value, m_rowsetSize);
+      return;
```

Comment:
Why do some cases in this switch return, and others break?

Reply:
The cases that break set the value `successfully_written` which is checked after the break. The ODBC then returns a warning if the attribute is not written successfully.